crates/red_knot/src/ast_ids.rs

```diff
@@ -0,0 +1,323 @@
+use std::fmt::Formatter;
+use std::marker::PhantomData;
+
+use rustc_hash::FxHashMap;
+
+use ruff_index::{Idx, IndexVec};
+use ruff_python_ast::visitor::preorder;
+use ruff_python_ast::visitor::preorder::{PreorderVisitor, TraversalSignal};
+use ruff_python_ast::{
+    AnyNodeRef, AstNode, ModModule, NodeKind, Parameter, Stmt, StmtAnnAssign, StmtAssign,
+    StmtAugAssign, StmtClassDef, StmtFunctionDef, StmtImport, StmtImportFrom, StmtTypeAlias,
+    TypeParam, TypeParamParamSpec, TypeParamTypeVar, TypeParamTypeVarTuple,
+};
+use ruff_text_size::{Ranged, TextRange};
+
+#[ruff_index::newtype_index]
+pub struct AstId;
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub struct FileAstId<N: HasAstId> {
+    ast_id: AstId,
+    _marker: PhantomData<fn() -> N>,
```

Comment:
Rust requires for structs that each type parameter is used by at least one field. Now, `AstId` isn't generic over `N`. 

We can work around this by using `PhantomData`. `PhantomData` has no runtime cost (it compiles down to a zero size type) and it's only purpose is to capture the type `N`. 

You can think of the pattern applied here as compile-time only generics similar to Java where the generic arguments are erased at runtime but we want them at compile time to catch typing errors (e.g. we want to prevent that you use a `AstId` for an `IfStmt` to load a `FunctionDef`).

Reply:
The thing I find a bit odd here is the `fn() -> N`; why is it not just `PhantomData<N>`? It seems like putting the typevar in return position like that might be intended to make `TypedAstId` contravariant in `N` rather than variant? But since Rust doesn't have struct subtyping, I'm not really sure what that would even mean; I thought Rust only really has variance for lifetimes.

Reply:
The reason is (and I was surprised by it) that `Phantom<Data>` made the type non-copyable, `NonEq` etc. 

I guess the alternative would have been to implement all these types manually https://users.rust-lang.org/t/how-to-copy-phantomdata-of-un-clone-able-types/82229

Reply:
Oh weird! So basically it's just that Rust knows how to implement `Eq` for `fn() -> N` but not necessarily for `N`?